src/Compilers/Core/Portable/Syntax/SyntaxTree.cs

```diff
@@ -343,6 +358,16 @@ internal Cci.DebugSourceInfo GetDebugSourceInfo()
         /// </summary>
         public abstract SyntaxTree WithFilePath(string path);
 
+        /// <summary>
+        /// Returns a new tree whose <see cref="DiagnosticOptions" /> are the specifed value and other properties are copied
+        /// from the current tree.
+        /// </summary>
+        /// <param name="options">
+        /// A mapping from diagnostic id to diagnostic reporting level. The diagnostic ID may be case-sensitive depending
+        /// on the language.
+        /// </param>
+        public abstract SyntaxTree WithDiagnosticOptions(ImmutableDictionary<string, ReportDiagnostic> options);
```

Comment:
i'm surprised our tooling didn't catch this as a breaking change.  Adding abstract methods to a publicly inheritable type is a problem, right?

Reply:
Good point, I can make this virtual and have it throw an exception.